File: snobot_sim_java/src/main/java/com/snobot/simulator/simulator_components/rev/RevManager.java
package com.snobot.simulator.simulator_components.rev;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import com.snobot.simulator.SensorActuatorRegistry;
import com.snobot.simulator.simulator_components.smart_sc.BaseCanSmartSpeedController;
import com.snobot.simulator.wrapper_accessors.DataAccessorFactory;

@SuppressWarnings({"PMD.ExcessiveMethodLength", "PMD.NcssCount", "PMD.CyclomaticComplexity"})
public class RevManager
{
    private static final Logger sLOGGER = LogManager.getLogger(RevManager.class);

    private final Map<String, Consumer<RevSpeedControllerSimWrapper>> mMotorControllerNormalCallbacks;
    private final Map<String, BiConsumer<RevSpeedControllerSimWrapper, Integer>> mMotorControllerSlottedCallbacks;


    public RevManager()
    {
        mMotorControllerNormalCallbacks = new HashMap<>();
        mMotorControllerNormalCallbacks.put("SetpointCommand", RevSpeedControllerSimWrapper::handleSetSetpointCommand);
        mMotorControllerNormalCallbacks.put("SetSensorType", RevSpeedControllerSimWrapper::handleSetSensorType);
        mMotorControllerNormalCallbacks.put("SetFeedbackDevice", RevSpeedControllerSimWrapper::handleSetFeedbackDevice);
        mMotorControllerNormalCallbacks.put("GetAppliedOutput", RevSpeedControllerSimWrapper::handleGetAppliedOutput);
        mMotorControllerNormalCallbacks.put("GetEncoderPosition", RevSpeedControllerSimWrapper::handleGetEncoderPosition);
        mMotorControllerNormalCallbacks.put("GetEncoderVelocity", RevSpeedControllerSimWrapper::handleGetEncoderVelocity);

        mMotorControllerSlottedCallbacks = new HashMap<>();
        mMotorControllerSlottedCallbacks.put("SetP", RevSpeedControllerSimWrapper::handleSetPGain);
        mMotorControllerSlottedCallbacks.put("SetI", RevSpeedControllerSimWrapper::handleSetIGain);
        mMotorControllerSlottedCallbacks.put("SetD", RevSpeedControllerSimWrapper::handleSetDGain);
        mMotorControllerSlottedCallbacks.put("SetFF", RevSpeedControllerSimWrapper::handleSetFFGain);
    }

    public void handleMessage(String aCallback, int aCanPort, ByteBuffer aData)
    {
        aData.order(ByteOrder.LITTLE_ENDIAN);

        Set<String> unsupportedFunctions = new HashSet<>();
        unsupportedFunctions.add("GetFault");

        switch (aCallback)
        {
        case "Create":
            createSim(aCanPort);
            break;
        case "SetFollow":
        {
            int followerID = aData.getInt();
            int leadId = followerID & 0x3F;

            BaseCanSmartSpeedController leadWrapper = getMotorControllerWrapper(leadId);
            RevSpeedControllerSimWrapper follower = getMotorControllerWrapper(aCanPort);

            sLOGGER.log(Level.INFO, "Setting SparkMax " + aCanPort + " to follow " + leadId);

            leadWrapper.addFollower(follower);

            break;
        }
        default:
            if (mMotorControllerNormalCallbacks.containsKey(aCallback))
            {
                RevSpeedControllerSimWrapper wrapper = getMotorControllerWrapper(aCanPort);
                mMotorControllerNormalCallbacks.get(aCallback).accept(wrapper);
            }
            else if (mMotorControllerSlottedCallbacks.containsKey(aCallback))
            {
                RevSpeedControllerSimWrapper wrapper = getMotorControllerWrapper(aCanPort);

                int slot = aData.getInt();
                mMotorControllerSlottedCallbacks.get(aCallback).accept(wrapper, slot);
            }
            else if (unsupportedFunctions.contains(aCallback))
            {
                sLOGGER.log(Level.DEBUG, "Unsupported option " + aCallback + "(" + aData.limit() + " bytes)");
            }
            else
            {
                sLOGGER.log(Level.WARN, "Unsupported option " + aCallback + "(" + aData.limit() + " bytes)");
            }
            break;
        }
    }

    protected void createSim(int aCanPort)
    {
        int simPort = aCanPort + BaseCanSmartSpeedController.sCAN_SC_OFFSET;
        if (!DataAccessorFactory.getInstance().getSpeedControllerAccessor().getPortList().contains(simPort))
        {
            sLOGGER.log(Level.WARN, "REV Motor Controller is being created dynamically instead of in the config file for port " + aCanPort);

            DataAccessorFactory.getInstance().getSpeedControllerAccessor().createSimulator(simPort, RevSpeedControllerSimWrapper.class.getName());
        }
        else if (!(SensorActuatorRegistry.get().getSpeedControllers().get(simPort) instanceof RevSpeedControllerSimWrapper))
        {
            sLOGGER.log(Level.FATAL, "A pre-registered motor controller of type " + SensorActuatorRegistry.get().getSpeedControllers().get(simPort).getClass()
                + " is the wrong type on port " + aCanPort);
            SensorActuatorRegistry.get().getSpeedControllers().remove(simPort);
            createSim(aCanPort);
            return;
        }
        SensorActuatorRegistry.get().getSpeedControllers().get(simPort).setInitialized(true);
    }

    public void reset()
    {

    }

    protected <WrapperType> WrapperType getMotorControllerWrapper(int aCanPort)
    {
        return (WrapperType) SensorActuatorRegistry.get().getSpeedControllers().get(aCanPort + 100);
    }
}

File: snobot_sim/src/main/native/cpp/SnobotSim/SimulatorComponents/RevWrappers/RevManager.cpp

#include "SnobotSim/SimulatorComponents/RevWrappers/RevManager.h"

#include <cstring>

#include "SnobotSim/GetSensorActuatorHelper.h"
#include "SnobotSim/Logging/SnobotLogger.h"
#include "SnobotSim/ModuleWrapper/Factories/FactoryContainer.h"
#include "SnobotSim/SimulatorComponents/RevWrappers/RevSpeedControllerSimWrapper.h"
#include "SnobotSim/SimulatorComponents/SmartSC/CanIdOffset.h"
#include "SnobotSim/SimulatorComponents/SmartSC/UnpackingUtils.h"

using namespace SnobotSim;

namespace
{

std::shared_ptr<RevpeedControllerSim> getMotorControllerWrapper(int aCanPort)
{
    auto rawWrapper = SensorActuatorRegistry::Get().GetISpeedControllerWrapper(aCanPort + CAN_SC_OFFSET, true);
    auto wrapper = std::dynamic_pointer_cast<RevpeedControllerSim>(rawWrapper);
    if (!wrapper)
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Wrapper not found for port " << aCanPort);
    }
    return wrapper;
    // return (CtreTalonSrxSpeedControllerSim) SensorActuatorRegistry.get().getSpeedControllers().get(aCanPort + CtreTalonSrxSpeedControllerSim.sCAN_SC_OFFSET);
}

void createSim(int aCanPort)
{
    int simPort = aCanPort + CAN_SC_OFFSET;

    auto wrapper = SensorActuatorRegistry::Get().GetISpeedControllerWrapper(aCanPort, false);
    if (!wrapper)
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_WARN, "Rev Motor Controller is being created dynamically instead of in the config file for port " << aCanPort);

        FactoryContainer::Get().GetSpeedControllerFactory()->Create(simPort, RevpeedControllerSim::TYPE);
        // SensorActuatorRegistry::Get().Register(simPort, std::shared_ptr<ISpeedControllerWrapper>(new CtreTalonSRXSpeedControllerSim(aCanPort)));
    }
    else if (!std::dynamic_pointer_cast<RevpeedControllerSim>(wrapper))
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Wrong motor type set up on " << aCanPort);
        return;
    }

    SensorActuatorRegistry::Get().GetISpeedControllerWrapper(simPort, true)->SetInitialized(true);
}

} // namespace


RevManager::RevManager()
{
    mMotorControllerNormalCallbacks["SetpointCommand"] = [](std::shared_ptr<RevpeedControllerSim> wrapper) { wrapper->handleSetSetpointCommand(); };
    mMotorControllerNormalCallbacks["SetSensorType"] = [](std::shared_ptr<RevpeedControllerSim> wrapper) { wrapper->handleSetSensorType(); };
    mMotorControllerNormalCallbacks["SetFeedbackDevice"] = [](std::shared_ptr<RevpeedControllerSim> wrapper) { wrapper->handleSetFeedbackDevice(); };
    mMotorControllerNormalCallbacks["GetAppliedOutput"] = [](std::shared_ptr<RevpeedControllerSim> wrapper) { wrapper->handleGetAppliedOutput(); };
    mMotorControllerNormalCallbacks["GetEncoderPosition"] = [](std::shared_ptr<RevpeedControllerSim> wrapper) { wrapper->handleGetEncoderPosition(); };
    mMotorControllerNormalCallbacks["GetEncoderVelocity"] = [](std::shared_ptr<RevpeedControllerSim> wrapper) { wrapper->handleGetEncoderVelocity(); };

    mMotorControllerSlottedCallbacks["SetP"] = [](std::shared_ptr<RevpeedControllerSim> wrapper, int slot) { wrapper->handleSetPGain(slot); };
    mMotorControllerSlottedCallbacks["SetI"] = [](std::shared_ptr<RevpeedControllerSim> wrapper, int slot) { wrapper->handleSetIGain(slot); };
    mMotorControllerSlottedCallbacks["SetD"] = [](std::shared_ptr<RevpeedControllerSim> wrapper, int slot) { wrapper->handleSetDGain(slot); };
    mMotorControllerSlottedCallbacks["SetFF"] = [](std::shared_ptr<RevpeedControllerSim> wrapper, int slot) { wrapper->handleSetFFGain(slot); };
}

void RevManager::Reset()
{
}

void RevManager::handleMessage(const std::string& aCallback, int aCanPort, uint8_t* aBuffer, int aLength)
{
    SNOBOT_LOG(SnobotLogging::LOG_LEVEL_INFO, "Getting Motor Controller Message " << aCallback << " on port " << aCanPort << "(" << aLength << " bytes)");

    if ("Create" == aCallback)
    {
        createSim(aCanPort);
    }
    else if ("SetFollow" == aCallback)
    {
        auto [followerID] = ExtractData<int>(aBuffer, aLength);
        int leadId = followerID & 0x3F;

        auto leadWrapper = getMotorControllerWrapper(leadId);
        auto follower = getMotorControllerWrapper(aCanPort);

        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_INFO, "Setting SparkMax " << aCanPort << " to follow " << leadId);

        leadWrapper->addFollower(follower);
    }
    else if (mMotorControllerNormalCallbacks.find(aCallback) != mMotorControllerNormalCallbacks.end())
    {
        auto wrapper = getMotorControllerWrapper(aCanPort);
        mMotorControllerNormalCallbacks[aCallback](wrapper);
    }
    else if (mMotorControllerSlottedCallbacks.find(aCallback) != mMotorControllerSlottedCallbacks.end())
    {
        auto wrapper = getMotorControllerWrapper(aCanPort);
        auto [slot, value] = ExtractData<int, double>(aBuffer, aLength);

        mMotorControllerSlottedCallbacks[aCallback](wrapper, slot);
    }
    else
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Unknown motor callback: " << aCallback);
    }
}
